File: crates/ldw/src-rs/ldw/model/parsed/visitor.rs
// Generated on 2024-11-13T09:33:43.385Z
use super::model::*;

pub trait Visitor {
    #[allow(unused_variables)]
    fn handle_model(self: &mut Self, node: &Box<Model>) {
        self.handle_fqn(&node.name);
        if let Some(value) = &node.parent_name {
            self.handle_fqn(&value);
        }
        for value in &node.values {
            self.handle_model_value(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_fqn(self: &mut Self, node: &Fqn) {}

    #[allow(unused_variables)]
    fn handle_definition(self: &mut Self, node: &Box<Definition>) {
        self.handle_type(&node.r#type);
    }

    #[allow(unused_variables)]
    fn handle_deletion(self: &mut Self, node: &Box<Deletion>) {}

    #[allow(unused_variables)]
    fn handle_member_modification(self: &mut Self, node: &Box<MemberModification>) {
        for value in &node.values {
            self.handle_member_modification_value(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_member_deletion(self: &mut Self, node: &Box<MemberDeletion>) {}

    #[allow(unused_variables)]
    fn handle_member_addition(self: &mut Self, node: &Box<MemberAddition>) {
        self.handle_member_addition_value(&node.value);
    }

    #[allow(unused_variables)]
    fn handle_type(self: &mut Self, node: &Type) {
        match node {
            Type::VoidType(value) => self.handle_void_type(value),
            Type::PrimitiveType(value) => self.handle_primitive_type(value),
            Type::EnumType(value) => self.handle_enum_type(value),
            Type::TypeWithStructure(value) => self.handle_type_with_structure(value),
            Type::NamedTypeReference(value) => self.handle_named_type_reference(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_void_type(self: &mut Self, node: &Box<VoidType>) {}

    #[allow(unused_variables)]
    fn handle_primitive_type(self: &mut Self, node: &PrimitiveType) {}

    #[allow(unused_variables)]
    fn handle_enum_type(self: &mut Self, node: &Box<EnumType>) {}

    #[allow(unused_variables)]
    fn handle_type_with_structure(self: &mut Self, node: &TypeWithStructure) {
        match node {
            TypeWithStructure::SumType(value) => self.handle_sum_type(value),
            TypeWithStructure::ProductType(value) => self.handle_product_type(value),
            TypeWithStructure::GenericType(value) => self.handle_generic_type(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_sum_type(self: &mut Self, node: &Box<SumType>) {
        for value in &node.members {
            self.handle_type(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_product_type(self: &mut Self, node: &Box<ProductType>) {
        for value in &node.members {
            self.handle_product_member(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_product_member(self: &mut Self, node: &Box<ProductMember>) {
        self.handle_type(&node.r#type);
    }

    #[allow(unused_variables)]
    fn handle_generic_type(self: &mut Self, node: &GenericType) {
        match node {
            GenericType::MapType(value) => self.handle_map_type(value),
            GenericType::SetType(value) => self.handle_set_type(value),
            GenericType::SequenceType(value) => self.handle_sequence_type(value),
            GenericType::OptionType(value) => self.handle_option_type(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_map_type(self: &mut Self, node: &Box<MapType>) {
        self.handle_type(&node.key_type);
        self.handle_type(&node.value_type);
    }

    #[allow(unused_variables)]
    fn handle_set_type(self: &mut Self, node: &Box<SetType>) {
        self.handle_type(&node.key_type);
    }

    #[allow(unused_variables)]
    fn handle_sequence_type(self: &mut Self, node: &Box<SequenceType>) {
        self.handle_type(&node.element_type);
    }

    #[allow(unused_variables)]
    fn handle_option_type(self: &mut Self, node: &Box<OptionType>) {
        self.handle_type(&node.r#type);
    }

    #[allow(unused_variables)]
    fn handle_named_type_reference(self: &mut Self, node: &Box<NamedTypeReference>) {
        self.handle_fqn(&node.fqn);
    }

    #[allow(unused_variables)]
    fn handle_model_value(self: &mut Self, node: &ModelValue) {
        match node {
            ModelValue::Deletion(value) => self.handle_deletion(value),
            ModelValue::MemberModification(value) => self.handle_member_modification(value),
            ModelValue::Definition(value) => self.handle_definition(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_member_modification_value(self: &mut Self, node: &MemberModificationValue) {
        match node {
            MemberModificationValue::MemberDeletion(value) => self.handle_member_deletion(value),
            MemberModificationValue::MemberAddition(value) => self.handle_member_addition(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_member_addition_value(self: &mut Self, node: &MemberAdditionValue) {
        match node {
            MemberAdditionValue::ProductMember(value) => self.handle_product_member(value),
            MemberAdditionValue::Type(value) => self.handle_type(value),
        }
    }
}

File: crates/codegen/ldw/src-ts/languages/ldw/model/parsed/visitor.ts
// Generated on 2024-10-15T17:18:16.268Z
import * as Model from './model';

export class Visitor {
    visitModel(node: Model.Model): void {
        this.visitFqn(node.name);
        if (node.parentName != undefined) {
            this.visitFqn(node.parentName);
        }
        node.values.forEach((x) => {
            switch (x.discriminator) {
                case Model.Discriminator.Deletion:
                    this.visitDeletion(x);
                    break;
                case Model.Discriminator.MemberModification:
                    this.visitMemberModification(x);
                    break;
                case Model.Discriminator.Definition:
                    this.visitDefinition(x);
                    break;
            }
        });
    }

    visitFqn(node: Model.Fqn): void {
        node.forEach((x) => {});
    }

    visitDefinition(node: Model.Definition): void {
        this.visitType(node.type);
    }

    visitDeletion(node: Model.Deletion): void {}

    visitMemberModification(node: Model.MemberModification): void {
        node.values.forEach((x) => {
            switch (x.discriminator) {
                case Model.Discriminator.MemberDeletion:
                    this.visitMemberDeletion(x);
                    break;
                case Model.Discriminator.MemberAddition:
                    this.visitMemberAddition(x);
                    break;
            }
        });
    }

    visitMemberDeletion(node: Model.MemberDeletion): void {}

    visitMemberAddition(node: Model.MemberAddition): void {
        switch (node.value.discriminator) {
            case Model.Discriminator.ProductMember:
                this.visitProductMember(node.value);
                break;
            case Model.Discriminator.VoidType:
            case Model.Discriminator.PrimitiveType:
            case Model.Discriminator.EnumType:
            case Model.Discriminator.SumType:
            case Model.Discriminator.ProductType:
            case Model.Discriminator.MapType:
            case Model.Discriminator.SetType:
            case Model.Discriminator.SequenceType:
            case Model.Discriminator.OptionType:
            case Model.Discriminator.NamedTypeReference:
                this.visitType(node.value);
                break;
        }
    }

    visitType(node: Model.Type): void {
        switch (node.discriminator) {
            case Model.Discriminator.VoidType:
                this.visitVoidType(node);
                break;
            case Model.Discriminator.PrimitiveType:
                this.visitPrimitiveType(node);
                break;
            case Model.Discriminator.EnumType:
                this.visitEnumType(node);
                break;
            case Model.Discriminator.SumType:
            case Model.Discriminator.ProductType:
            case Model.Discriminator.MapType:
            case Model.Discriminator.SetType:
            case Model.Discriminator.SequenceType:
            case Model.Discriminator.OptionType:
                this.visitTypeWithStructure(node);
                break;
            case Model.Discriminator.NamedTypeReference:
                this.visitNamedTypeReference(node);
                break;
        }
    }

    visitVoidType(node: Model.VoidType): void {}

    visitPrimitiveType(node: Model.PrimitiveType): void {}

    visitEnumType(node: Model.EnumType): void {
        node.members.forEach((x) => {});
    }

    visitTypeWithStructure(node: Model.TypeWithStructure): void {
        switch (node.discriminator) {
            case Model.Discriminator.SumType:
                this.visitSumType(node);
                break;
            case Model.Discriminator.ProductType:
                this.visitProductType(node);
                break;
            case Model.Discriminator.MapType:
            case Model.Discriminator.SetType:
            case Model.Discriminator.SequenceType:
            case Model.Discriminator.OptionType:
                this.visitGenericType(node);
                break;
        }
    }

    visitSumType(node: Model.SumType): void {
        node.members.forEach((x) => {
            this.visitType(x);
        });
    }

    visitProductType(node: Model.ProductType): void {
        node.members.forEach((x) => {
            this.visitProductMember(x);
        });
    }

    visitProductMember(node: Model.ProductMember): void {
        this.visitType(node.type);
    }

    visitGenericType(node: Model.GenericType): void {
        switch (node.discriminator) {
            case Model.Discriminator.MapType:
                this.visitMapType(node);
                break;
            case Model.Discriminator.SetType:
                this.visitSetType(node);
                break;
            case Model.Discriminator.SequenceType:
                this.visitSequenceType(node);
                break;
            case Model.Discriminator.OptionType:
                this.visitOptionType(node);
                break;
        }
    }

    visitMapType(node: Model.MapType): void {
        this.visitType(node.keyType);
        this.visitType(node.valueType);
    }

    visitSetType(node: Model.SetType): void {
        this.visitType(node.keyType);
    }

    visitSequenceType(node: Model.SequenceType): void {
        this.visitType(node.elementType);
    }

    visitOptionType(node: Model.OptionType): void {
        this.visitType(node.type);
    }

    visitNamedTypeReference(node: Model.NamedTypeReference): void {
        this.visitFqn(node.fqn);
    }

    visitTrivia(node: Model.Trivia): void {
        switch (node.discriminator) {
            case Model.Discriminator.LineComment:
                this.visitLineComment(node);
                break;
            case Model.Discriminator.BlockComment:
                this.visitBlockComment(node);
                break;
            case Model.Discriminator.Whitespace:
                this.visitWhitespace(node);
                break;
        }
    }

    visitWhitespace(node: Model.Whitespace): void {}

    visitLineComment(node: Model.LineComment): void {}

    visitBlockComment(node: Model.BlockComment): void {}
}
